example/snake/model.hpp

```diff
@@ -0,0 +1,73 @@
+//
+// lager - library for functional interactive c++ programs
+// Copyright (C) 2019 Carl Bussey
+//
+// This file is part of lager.
+//
+// lager is free software: you can redistribute it and/or modify
+// it under the terms of the MIT License, as detailed in the LICENSE
+// file located at the root of this source code distribution,
+// or here: <https://github.com/arximboldi/lager/blob/master/LICENSE>
+//
+
+#pragma once
+
+#include <cstdlib>
+#include <ctime>
+#include <utility>
+#include <vector>
+
+namespace sn {
+
+using point_t = std::pair<int, int>;
+
+inline int x(point_t p) { return p.first; }
+
+inline int y(point_t p) { return p.second; }
+
+enum class direction
+{
+    left,
+    up,
+    right,
+    down
+};
+
+struct snake_t
+{
+    using body_t = std::vector<point_t>;
+    body_t body{};
+    direction dir{};
+};
+
+struct game_t
+{
+    snake_t snake;
+    point_t apple_pos{};
+    bool over;
+
+    static constexpr int width{25};
+    static constexpr int height{25};
+};
+
+inline point_t random_apple_pos()
+{
+    static const auto seeder = [] {
+        std::srand(std::time(nullptr));
```

Comment:
:warning: `std::time` and `std::srand` are not pure functions! Do not use in the reducer!

You have two options:
1. use an effect to compute the random number, then return it back into the game with an action, or 
2. put the pseudo-random sequence as part of the model, for example using the stuff from `<random>`. You can get a entrophised seed before creating the store.

Reply:
Thanks for pointing this out. In case anybody else is reading this, I thought it would be worth adding your elaboration here (hope you don't see this as a betrayal of your trust by sharing a private conversation!):

> Testability is one reason.
> But yeah, the idea is that when you replay the same sequence of actions you get the same result
> for example, the time travelling debugger now stores previous states
> but in theory it should not need to
> it can just save snapshots from time to time and interpolate replaying the actions in between